### ISSUE TITLE
"terragrunt apply" fails with AWS acm module (terraform-aws-modules/acm/aws?version=5.1.1)

### ISSUE THREAD

**Issue by robertzakrocki** [gruntwork-io/terragrunt#3471]
## Describe the bug

I am trying to use *terraform-aws-modules/acm/aws?version=5.1.1* module to create an ssl certificate,

When applying the terragrunt plan to create the ssl certificate and to validate it (using the DNS option in a module), the application of the plan fails with the following error:

Please note I've masked the domain name (with subdomain.domain.ac.uk) and Route53 Zone (with ZXXXXXXXXXXXXXXXXXXX) in the error output.

aws_route53_record.validation[0]: Creating...

╷
│ Error: reading Route 53 Hosted Zone ({"subdomain.domain.ac.uk":"ZXXXXXXXXXXXXXXXXXXX"}): operation error Route 53: GetHostedZone, https response error StatusCode: 400, RequestID: f2fa3d59-1fa2-4b48-8d44-af021930cabc, InvalidInput: 1 validation error detected: Value '{"subdomain.domain.ac.uk":"ZXXXXXXXXXXXXXXXXXXX"}' at 'resourceId' failed to satisfy constraint: Member must have length less than or equal to 32
│
│   with aws_route53_record.validation[0],
│   on main.tf line 47, in resource "aws_route53_record" "validation":
│   47: resource "aws_route53_record" "validation" {
│
╵

exit status 1
11:27:50.435 ERROR  terraform invocation failed in ./.terragrunt-cache/_4fe1fTQfRr5Jf34waTt_qx-bGc/MaHcPPxgf1hzgfp4nD8LyUXfWtQ stderr=
╷
│ Error: reading Route 53 Hosted Zone ({"subdomain.domain.ac.uk":"ZXXXXXXXXXXXXXXXXXXX"}): operation error Route 53: GetHostedZone, https response error StatusCode: 400, RequestID: f2fa3d59-1fa2-4b48-8d44-af021930cabc, InvalidInput: 1 validation error detected: Value '{"subdomain.domain.ac.uk":"ZXXXXXXXXXXXXXXXXXXX"}' at 'resourceId' failed to satisfy constraint: Member must have length less than or equal to 32
│
│   with aws_route53_record.validation[0],
│   on main.tf line 47, in resource "aws_route53_record" "validation":
│   47: resource "aws_route53_record" "validation" {
│
╵

11:27:50.437 ERROR  1 error occurred:
	* [./.terragrunt-cache/_4fe1fTQfRr5Jf34waTt_qx-bGc/MaHcPPxgf1hzgfp4nD8LyUXfWtQ] exit status 1

## Steps To Reproduce

Try applying the code below, please note that terragrunt plan, plans it okay but the application of the plan fails. The certificate gets created but the cname records (for dns validations) aren't.

```hcl

terraform {
  source = "tfr:///terraform-aws-modules/acm/aws?version=5.1.1"
}

inputs = {

  domain_name        = "subdomain.domain.ac.uk"
  zone_id                  = "ZXXXXXXXXXXXXXXXXXXX"
  validation_method  = "DNS"
  
  subject_alternative_names = [
    "*.subdomain.domain.ac.uk"
  ]

  tags = {
    ManagedBy = "Terraform/Terragrunt"
  }
}
```

## Expected behavior

* SSL certificate created
* SSL certificate validated (by creation cname record in an existing Route53 Zone)

When using just terraform this works fine. This is the code that I have used with terraform on it's own:

```hcl


terraform {
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "~> 4.16"
    }
  }

}

provider "aws" {
  region  = "eu-west-1"
}

module "acm" {
  source  = "terraform-aws-modules/acm/aws"
  version = "5.1.1"

  domain_name  = "subdomain.domain.ac.uk"
  zone_id      = "ZXXXXXXXXXXXXXXXXXXX"

  validation_method = "DNS"

  subject_alternative_names = [
    "*.subdomain.domain.ac.uk",
  ]


  tags = {
    ManagedBy = "Terraform/Terragrunt"
  }
}
``` 

And the console output from the terraform:

```
Plan: 3 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

module.acm.aws_acm_certificate.this[0]: Creating...
module.acm.aws_acm_certificate.this[0]: Creation complete after 6s [id=arn:aws:acm:eu-west-1:012345678901:certificate/bc8cdf1f-4fe3-4ccd-a6bf-85bce0c7eb94]
module.acm.aws_route53_record.validation[0]: Creating...
module.acm.aws_route53_record.validation[0]: Still creating... [10s elapsed]
module.acm.aws_route53_record.validation[0]: Still creating... [20s elapsed]
module.acm.aws_route53_record.validation[0]: Still creating... [30s elapsed]
module.acm.aws_route53_record.validation[0]: Creation complete after 37s [id=ZXXXXXXXXXXXXXXXXXX__6db6c4caaa02e26171f657deeff24fe6.subdomain.domain.ac.uk._CNAME]
module.acm.aws_acm_certificate_validation.this[0]: Creating...
module.acm.aws_acm_certificate_validation.this[0]: Creation complete after 0s [id=2024-10-12 11:18:08.431 +0000 UTC]

Apply complete! Resources: 3 added, 0 changed, 0 destroyed.
```
## Nice to haves

- [ x] Terminal output
- [ x] Screenshots
![Screenshot 2024-10-12 at 12 32 44](https://github.com/user-attachments/assets/6e0f65c8-5a8b-44d6-97e2-64b24d8db8ee)

## Versions

- Terragrunt version: terragrunt version 0.67.16
- OpenTofu/Terraform version: Terraform v1.9.7
- Environment details (Ubuntu 20.04, Windows 10, etc.): MacOS 14.7

## Additional context

* hashicorp/aws v5.71.0
* Tried deleting terragrunt cache

**Comment by yhakbar** [comment]
Closing out this issue, as the root cause is an error thrown by the AWS Provider, not Terragrunt:

```
│ Error: reading Route 53 Hosted Zone ({"subdomain.domain.ac.uk":"ZXXXXXXXXXXXXXXXXXXX"}): operation error Route 53: GetHostedZone, https response error StatusCode: 400, RequestID: f2fa3d59-1fa2-4b48-8d44-af021930cabc, InvalidInput: 1 validation error detected: Value '{"subdomain.domain.ac.uk":"ZXXXXXXXXXXXXXXXXXXX"}' at 'resourceId' failed to satisfy constraint: Member must have length less than or equal to 32
```

Ensure that the values you are setting as inputs are correct, following @denis256 's advice.

**Comment by yhakbar** [comment]
Hey @robertzakrocki ,

I'd like confirmation that this is, in-fact a Terragrunt issue before re-opening, as it does look to be an issue with how inputs are being passed down to Terraform, which passes values to the AWS provider.

I'm happy to help you debug, though.

Note the output when I run the command with `--terragrunt-debug` in my local copy of your `terragrunt.hcl`:

```json
{
  "domain_name": "subdomain.domain.ac.uk",
  "subject_alternative_names": [
    "*.subdomain.domain.ac.uk"
  ],
  "tags": {
    "ManagedBy": "Terraform/Terragrunt"
  },
  "validation_method": "DNS",
  "zone_id": "ZXXXXXXXXXXXXXXXXXXX"
}
```

That's different than what you're seeing in your `terragrunt-debug.tfvars.json` file. Note this part of your file:

```json
  "zone_id": {
    "subdomain.domain.ac.uk": "ZXXXXXXXXXXXXXXXXXXX"
  }
```

This makes the error message you're seeing from the AWS provider make sense, as it's not seeing `ZXXXXXXXXXXXXXXXXXXX` as the `zone_id`, but `{"subdomain.domain.ac.uk":"ZXXXXXXXXXXXXXXXXXXX"}`:

```
Error: reading Route 53 Hosted Zone ({"subdomain.domain.ac.uk":"ZXXXXXXXXXXXXXXXXXXX"})
```

Are you sure you didn't edit the `terragrunt.hcl` file and forget to save or something?

**Comment by robertzakrocki** [comment]
Hey @yhakbar ,

Apologies, when you've pointed out the difference between your test and my test I realised where the issue is.

To simplify the code when I initially posted the issue I replaced variables with  the masked values. The variable I used for zone_id was an output from another stack

`zone_id = dependency.dns.outputs.route53_zone_zone_id `

I expected to just get zone_id in a form of a string (ZXXXXXXXXX) but evidently that isn't the case. I have used this module: https://registry.terraform.io/modules/terraform-aws-modules/route53/aws/latest/submodules/zones?tab=outputs. 

I have just tested the output from the dns stack and I get the output in this form:

 route53_zone_zone_id = {
"subdomain.domain.ac.uk" = "ZXXXXXXXXXXXXXXX"
} 

Which explains why the terragrunt plan was failing
